src/vea/vea_api.c

```diff
@@ -275,6 +314,146 @@ aging_flush(struct vea_space_info *vsi, bool force, uint32_t nr_flush, uint32_t
 	}
 }
 
+struct reclaim_commit_cb_arg {
+	struct vea_space_info	*rca_vsi;
+	struct vea_free_entry	 rca_vfe;
+};
+
+static void
+reclaim_commit_cb(void *data, bool noop)
+{
+	struct reclaim_commit_cb_arg	*rca = data;
+	d_iov_t				 key;
+	int				 rc;
+
+	/* Transaction aborted, only need to free callback arg */
+	if (noop)
+		goto free;
+
+	/*
+	 * Even in-memory bitmap failed to remove from tree, it is ok
+	 * because this bitmap chunk has been removed from allocation LRU list.
+	 */
+	d_iov_set(&key, &rca->rca_vfe.vfe_ext.vfe_blk_off,
+		  sizeof(rca->rca_vfe.vfe_ext.vfe_blk_off));
+	rc = dbtree_delete(rca->rca_vsi->vsi_bitmap_btr, BTR_PROBE_EQ, &key, NULL);
+	if (rc) {
+		D_ERROR("Remove ["DF_U64", %u] from bitmap tree "
+			"error: "DF_RC"\n", rca->rca_vfe.vfe_ext.vfe_blk_off,
+			rca->rca_vfe.vfe_ext.vfe_blk_cnt, DP_RC(rc));
+		goto free;
+	}
+	dec_stats(rca->rca_vsi, STAT_FRAGS_BITMAP, 1);
+	/*
+	 * Aggregated free will be executed on outermost transaction
+	 * commit.
+	 *
+	 * If it fails, the freed space on persistent free tree won't
+	 * be added in in-memory free tree, hence the space won't be
+	 * visible for allocation until the tree sync up on next server
+	 * restart. Such temporary space leak is tolerable, what we must
+	 * avoid is the contrary case: in-memory tree update succeeds
+	 * but persistent tree update fails, which risks data corruption.
+	 */
+	rc = aggregated_free(rca->rca_vsi, &rca->rca_vfe);
+
+	D_CDEBUG(rc, DLOG_ERR, DB_IO, "Aggregated free on vsi:%p rc %d\n",
+		 rca->rca_vsi, rc);
+free:
+	D_FREE(rca);
+}
+
+static bool
+is_bitmap_empty(uint64_t *bitmap, int bitmap_sz)
+{
+	int i;
+
+	for (i = 0; i < bitmap_sz; i++)
+		if (bitmap[i])
+			return false;
+
+	return true;
+}
+
+static int
+reclaim_unused_bitmap(struct vea_space_info *vsi, uint32_t nr_reclaim, uint32_t *nr_reclaimed)
+{
+	int				 i;
+	struct vea_bitmap_entry		*bitmap_entry, *tmp_entry;
+	struct vea_free_bitmap		*vfb;
+	d_iov_t				 key;
+	int				 rc = 0;
+	struct reclaim_commit_cb_arg	*rca;
+	struct umem_instance		*umem = vsi->vsi_umem;
+	int				 nr = 0;
+	uint64_t			 blk_off;
+	uint32_t			 blk_cnt;
+
+	for (i = 0; i < VEA_MAX_BITMAP_CLASS; i++) {
+		d_list_for_each_entry_safe(bitmap_entry, tmp_entry,
+				&vsi->vsi_class.vfc_bitmap_lru[i], vbe_link) {
+			vfb = &bitmap_entry->vbe_bitmap;
+			D_ASSERT(vfb->vfb_class == i + 1);
+			if (!is_bitmap_empty(vfb->vfb_bitmaps, vfb->vfb_bitmap_sz))
```

Comment:
OK, that means we need check bitmap chunks for each free, and if it empty, we put it on empty list, do you think it ok for doing this or we'd better maintain a free counter of bitmap to avoid calculating it every time?

Reply:
oh, maybe it is fine, as jeff mentioned we have some builtin helper .